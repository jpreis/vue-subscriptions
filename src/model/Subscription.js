export default class Subscription {
  static nextId = 0;

  constructor(name, costs, active = true) {
    this.id = Subscription.nextId++;
    this.name = name;
    this.costs = costs;
    this.active = active;
  }
}
