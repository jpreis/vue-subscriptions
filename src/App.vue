<template>
  <div id="app">
    <h1>My subscriptions</h1>

    <form @submit.prevent="addSubscription">
      <input class="subscription-input" type="text" placeholder="subscription (e.g. Netflix)" v-model="newSubscriptionName">
      <input class="subscription-input" type="text" placeholder="monthly costs (e.g. 9.99)" v-model="newSubscriptionCosts">

      <button type="submit">Add {{ newSubscriptionName }}</button>
    </form>

    <SubscriptionStats :subscriptions="subscriptions"></SubscriptionStats>

    <div class="subscription-cards">
      <SubscriptionCard v-for="subscription in subscriptions " :subscription="subscription" :key="subscription.id" @toggle-active="toggleActive(subscription)"></SubscriptionCard>
    </div>

  </div>
</template>

<script>
import Subscription from "@/model/Subscription";
import SubscriptionStats from "@/components/SubscriptionStats.vue";
import SubscriptionCard from "@/components/SubscriptionCard.vue";

export default {
  data: function() {
    return {
      newSubscriptionName: "",
      newSubscriptionCosts: "",
      subscriptions: []
    };
  },
  methods: {
    addSubscription() {
      let name = this.newSubscriptionName;
      let costs = parseFloat(this.newSubscriptionCosts);
      if (name && costs >= 0) {
        this.subscriptions.push(new Subscription(name, costs));
        this.newSubscriptionName = "";
        this.newSubscriptionCosts = "";
      }
    },
    toggleActive(subscription) {
      subscription.active = !subscription.active;
    }
  },
  components: {
    SubscriptionStats,
    SubscriptionCard
  }
};
</script>

<style lang="scss">
@import url("https://fonts.googleapis.com/css?family=Overpass:400,900");

:root {
  --vue-green: #56b889;
  --vue-gray: #445465;
  --vue-white: #f8f8f8;
  --font-family: "Overpass", sans-serif;
}

html,
body {
  margin: 0;
  padding: 0;
}

html {
  font-size: 1.25em;
}

body {
  background-color: var(--vue-white);
  font-family: var(--font-family);
  color: #333;
}

#app {
  max-width: 900px;
  margin: 0 auto;
  padding: 4em 1em;
}

.subscription-cards {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
  grid-gap: 2.5rem;
  grid-auto-rows: 1fr;
  margin-top: 2rem;
  padding-top: 2rem;
  padding-bottom: 2rem;
}

.subscription-input {
  --border-color: var(--vue-gray);
  background-color: rgba(0, 0, 0, 0.025);
  border: none;
  font-size: 1rem;
  font-family: var(--font-family);
  border-bottom: 2px solid var(--border-color);
  margin-right: 1em;
  padding: 0.2em 0.4em;
  width: 300px;

  &::placeholder {
    color: rgba(0, 0, 0, 0.3);
  }

  &:focus {
    outline: none;
    --border-color: var(--vue-green);
  }
}

button {
  --button-background: var(--vue-green);
  --button-border: transparent;
  --button-color: white;

  font-family: var(--font-family);
  font-size: 1rem;
  padding: 0.2em 1em;
  color: var(--button-color);
  background-color: var(--button-background);
  border-radius: 5px;
  border: 2px solid var(--button-border);
  transition: all 0.3s;
  box-shadow: 0 7px 14px rgba(0, 0, 0, 0.075), 0 5px 5px rgba(0, 0, 0, 0.06);

  &:hover,
  &:active {
    --button-background: white;
    --button-border: var(--vue-green);
    --button-color: var(--vue-green);
    cursor: pointer;
  }

  &:focus {
    outline: none;
  }
}

p {
  vertical-align: center;
}
</style>
