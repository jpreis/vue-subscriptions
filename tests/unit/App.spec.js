import { shallowMount, mount } from "@vue/test-utils";
import App from "@/App.vue";
import Subscription from "@/model/Subscription";
import Vue from "vue";

describe("App.vue", () => {
  let wrapper;
  let data;

  beforeEach(() => {
    wrapper = shallowMount(App, {});
    data = wrapper.vm.$data;
  });

  it("has the expected HTML structure", () => {
    wrapper = mount(App, {
      stubs: {
        SubscriptionCard: "<p class='subscriptionCard'></p>"
      }
    });

    expect(wrapper.element).toMatchSnapshot();

    // add a subscription
    // ! directily adding the subscription in data will not trigger a re-render at the moment
    // * see: https://github.com/vuejs/vue-test-utils/issues/605
    wrapper.setData({
      newSubscriptionName: "Testsubscription",
      newSubscriptionCosts: "10.99"
    });
    wrapper.vm.addSubscription();

    expect(wrapper.element).toMatchSnapshot();
  });

  describe("adding subscriptions", () => {
    it("adds a new subscriptions", () => {
      wrapper.setData({
        newSubscriptionName: "Testsubscription",
        newSubscriptionCosts: "10.99"
      });
      wrapper.vm.addSubscription();
      expect(data.subscriptions.length).toBe(1);
    });

    it("does not add a new subscription if the user input is invalid", () => {
      // no subscription name entered
      wrapper.setData({
        newSubscriptionName: "",
        newSubscriptionCosts: "10.99"
      });

      wrapper.vm.addSubscription();
      expect(data.subscriptions.length).toBe(0);

      // invalid costs (negative)
      wrapper.setData({
        newSubscriptionName: "Testsubscription",
        newSubscriptionCosts: "-1"
      });

      wrapper.vm.addSubscription();
      expect(data.subscriptions.length).toBe(0);

      // invalid costs (not a number)
      wrapper.setData({
        newSubscriptionCosts: "foo"
      });

      wrapper.vm.addSubscription();
      expect(data.subscriptions.length).toBe(0);
    });
  });
});
