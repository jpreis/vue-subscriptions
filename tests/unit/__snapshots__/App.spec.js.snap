// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue has the expected HTML structure 1`] = `
<div
  id="app"
>
  <h1>
    My subscriptions
  </h1>
   
  <form>
    <input
      class="subscription-input"
      placeholder="subscription (e.g. Netflix)"
      type="text"
    />
     
    <input
      class="subscription-input"
      placeholder="monthly costs (e.g. 9.99)"
      type="text"
    />
     
    <button
      type="submit"
    >
      Add 
    </button>
  </form>
   
  <p
    class="subscription-stats"
  >
    
    No subscriptions yet...
  
  </p>
   
  <div
    class="subscription-cards"
  />
</div>
`;

exports[`App.vue has the expected HTML structure 2`] = `
<div
  id="app"
>
  <h1>
    My subscriptions
  </h1>
   
  <form>
    <input
      class="subscription-input"
      placeholder="subscription (e.g. Netflix)"
      type="text"
    />
     
    <input
      class="subscription-input"
      placeholder="monthly costs (e.g. 9.99)"
      type="text"
    />
     
    <button
      type="submit"
    >
      Add 
    </button>
  </form>
   
  <p
    class="subscription-stats"
  >
    
    I currently have
    
    <strong>
      1 active subscription
    </strong>
    . 
    <br />
     That's
    
    <strong>
      $10.99 a month 
    </strong>
     and
    
    <strong>
      $131.88 a year
    </strong>
    . 
  </p>
   
  <div
    class="subscription-cards"
  >
    <p
      class="subscriptionCard"
    />
  </div>
</div>
`;
